README

Tony
====

Tony is a program for computer-aided melody annotation. It has a
graphical interface based on the Sonic Visualiser libraries, and uses
the pYIN Vamp plugin to extract pitch track and notes from monophonic
audio.


Features
========

 * robust monophonic pitch track extraction (using pYIN)
 * note track extraction
 * facility to manually adjust pitch track and note track
 * facility to audition pitch and note track
 * note pitch automatically snaps to pitch track
 * import/export of pitch track and note track


Authors, Citation, License and Use
==================================

Tony was developed at Queen Mary, University of London in
collaboration with New York University.

Code copyright 2005-2015 Chris Cannam, Queen Mary University of
London, and the Tony project authors: Matthias Mauch, George Fazekas,
Justin Salamon, and Rachel Bittner, except where indicated in the
individual source files. Thanks also to Simon Dixon and Juan Bello.

If you make use of this software for any public or commercial purpose,
we ask you to kindly mention the authors and Queen Mary, University of
London in your user-visible documentation. We're very happy to see
this sort of use but would much appreciate being credited, separately
from the requirements of the software license itself (see below).

If you make use of this software for academic purposes, please cite
one of the publications indicated on the Publications page:
https://code.soundsoftware.ac.uk/publications?project_id=tony

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A ARTICULAR PURPOSE. See the GNU
General Public License for more details. You should have received a
copy of the GNU General Public License along with this program. If
not, see http://www.gnu.org/licenses/.



CHANGELOG

Changes in Tony 2.1.1 (10 Dec 2019) since the previous release 2.1:

 - Fix incorrect note numbering in display

 - Fix incorrect state of "Penalise Soft Pitches" option in Analysis
   menu. Previously this was displayed as if enabled, but the option
   was not actually active when performing the analysis, until the
   option was toggled or "Reset Options to Default" was used

 - Fix failure to open playback device when using macOS Catalina with
   certain audio devices

 - Remove dependency on JACK in Linux deb package: the application can
   use it if available, but it isn't required

 - Fix failure to make installed executables executable when
   installing from the source package


Changes in Tony 2.1 (25 Oct 2019) since the previous release 2.0:

 - Fix problems reopening session files

 - Add support for reading the Opus audio codec on all platforms

 - Add support for WMA and AAC formats on 64-bit Windows. (AAC was
   already supported on macOS.) Note that AAC support on Windows does
   not support "gapless" import, i.e. removing the encoder gaps at
   start and end of the stream

 - Fix sluggishness in starting and stopping playback when using
   PulseAudio v13.0+

 - Update platform builds with newer libraries and updated platform
   support

Changes in Tony 2.0 (23 Oct 2015) since the previous release 1.0:

 - Tony now has a Record button, allowing you to record audio from the
   microphone and analyse it

 - Graphical rendering and icons now properly support hi-dpi/Retina
   displays

 - The playback speed dial uses a more sensible scale

Changes in Tony 1.0 (01 April 2015) since the previous release 0.6:

 - The underlying pYin analysis plugin has been updated; more analysis
   options are available

 - Notes can be dragged up and down using the mouse in edit mode. When
   you drag a note, the pitch track underneath it snaps to the nearest
   spectral peaks (if any are found). Releasing the note makes the
   note snap back to the (possibly now moved) underlying pitch values.

 - Exported pitch track CSV files now contain one row for every
   processing frame in the file, and those with no identified pitch
   have the pitch recorded as 0. This is useful for some evaluation
   and post-processing tools, and it's easier to filter out the 0
   values if you don't need them than add them afterwards if you do

 - Exported note track CSV files now omit the (meaningless, in this
   context) velocity column

 - Level and pan for a layer (audio, pitch track, notes) are now shown
   in a little matrix widget, which is quicker to read than the
   previous pair of rotary dials; click on it to pop up an editable
   version

 - Horizontal scrolling using a Mac trackpad now works properly

 - Tony now finds plugins in its own installation directory by default
   (can be overridden with TONY_VAMP_PATH environment variable)

 - Bug fixes

Changes in Tony 0.6 (16 June 2014) since the previous release 0.5:

 - The play pointer no longer tracks to the centre of the main pane
   when dragging, and is now always visible. Play pointer / playhead
   behaviour is generally expected to be much more sensible and
   intuitive in this release than in prior ones

 - Many keyboard shortcuts have been added and improved, including the
   shortcuts for incremental note-by-note selection and moving the
   play pointer

 - Audio files are now implicitly normalised to peak == 1.0 when
   loading

 - Tony now has an Analysis menu which contains various configuration
   settings that affect pitch analysis

 - There is now an option to save the session directly to the same
   directory as the original audio file, for rapid saving when
   annotating many audio files in sequence

 - The extents of the "active" note are now more clearly shown when
   mousing over the note track

 - There is a new function to delete the notes (without deleting pitch
   measurements) within the selection

 - Numerous bug fixes

Changes in Tony 0.5 (03 April 2014) since the previous release 0.4:

 - The start and end of the recording are now shaded in grey so you
   can see where they are

 - There are new save/open session functions on the File menu; the
   session file extension is .ton, and save/open session should work

 - The note edit tool (which was largely broken in 0.4 after lots of
   work on pitch track editing) now works again

 - New note editing menu actions: split note, merge notes, form note
   from selection, snap notes to pitch track

 - Editing the pitch track (when notes are present) causes the notes
   to be automatically updated to the new pitch track

 - The Key and Mouse Reference help window has been updated so as to
   show Tony-specific actions properly

 - The undo/redo command bundling has been tidied up and some related
   crashes fixed
	
Changes in Tony 0.4 (12 March 2014) since the previous release 0.3:

 - The main pitch tracker (pYIN) now uses a non-FFT Yin method in
   order to produce unbiased timestamps

 - Pitch estimation within a constrained frequency range now uses a
   harmonic spectrum plugin (CHP) that tracks peaks within the range
   in a way more in line with user expectations than pYIN

 - The layer show/hide controls have been merged into the layer type
   icons, saving toolbar space

 - Adjustment of the displayed frequency range has been crudely added

 - Fix crash on loading a second file and inability to export pitch
   track after using selection strip

Changes in Tony 0.3 (06 March 2014) since the previous release 0.2:

 - Key shortcuts for editing have been revised to be more consistent
   on OS/X and across different Qt versions

 - Reduce visual clutter in selection areas

 - Fixes to OS/X deployment

Changes in Tony 0.2 (05 March 2014) since the previous release 0.1:


main/main.cpp
/* -*- c-basic-offset: 4 indent-tabs-mode: nil -*-  vi:set ts=8 sts=4 sw=4: */

/*
    Tony
    An intonation analysis and annotation tool
    Centre for Digital Music, Queen Mary, University of London.
    This file copyright 2006-2012 Chris Cannam and QMUL.
    
    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.  See the file
    COPYING included with this distribution for more information.
*/

#include "MainWindow.h"

#include "system/System.h"
#include "system/Init.h"
#include "base/TempDirectory.h"
#include "base/PropertyContainer.h"
#include "base/Preferences.h"
#include "widgets/TipDialog.h"
#include "widgets/InteractiveFileFinder.h"
#include "transform/TransformFactory.h"
#include "svcore/plugin/PluginScan.h"

#include <QMetaType>
#include <QApplication>
#include <QScreen>
#include <QMessageBox>
#include <QTranslator>
#include <QLocale>
#include <QSettings>
#include <QIcon>
#include <QSessionManager>
#include <QSplashScreen>
#include <QFileOpenEvent>
#include <QDir>

#include <iostream>
#include <signal.h>
#include <cstdlib>

#include "../version.h"

#include <vamp-hostsdk/PluginHostAdapter.h>

static QMutex cleanupMutex;
static bool cleanedUp = false;

static void
signalHandler(int /* signal */)
{
    // Avoid this happening more than once across threads

    cerr << "signalHandler: cleaning up and exiting" << endl;

    if (cleanupMutex.tryLock(5000)) {
        if (!cleanedUp) {
            TempDirectory::getInstance()->cleanup();
            cleanedUp = true;
        }
        cleanupMutex.unlock();
    }
    
    exit(0);
}

class TonyApplication : public QApplication
{
public:
    TonyApplication(int &argc, char **argv) :
        QApplication(argc, argv),
        m_mainWindow(0),
        m_readyForFiles(false)
    {
        // tidier without, I reckon
        setAttribute(Qt::AA_DontShowIconsInMenus);
    }
    virtual ~TonyApplication() {
    }

    void setMainWindow(MainWindow *mw) { m_mainWindow = mw; }
    void releaseMainWindow() { m_mainWindow = 0; }

    virtual void commitData(QSessionManager &manager) {
        if (!m_mainWindow) return;
        bool mayAskUser = manager.allowsInteraction();
        bool success = m_mainWindow->commitData(mayAskUser);
        manager.release();
        if (!success) manager.cancel();
    }

    void readyForFiles() {
        m_readyForFiles = true;
    }

    void handleFilepathArgument(QString path, QSplashScreen *splash);

    void handleQueuedPaths(QSplashScreen *splash) {
        foreach (QString f, m_filepathQueue) {
            handleFilepathArgument(f, splash);
        }
    }
    
protected:
    MainWindow *m_mainWindow;
    
    bool m_readyForFiles;
    QStringList m_filepathQueue;

    virtual bool event(QEvent *event) {

        if (event->type() == QEvent::FileOpen) {
            QString path = static_cast<QFileOpenEvent *>(event)->file();
            if (m_readyForFiles) {
                handleFilepathArgument(path, NULL);
            } else {
                m_filepathQueue.append(path);
            }
            return true;
        } else {
            return QApplication::event(event);
        }
    }
};

static QString
getEnvQStr(QString variable)
{
#ifdef Q_OS_WIN32
    std::wstring wvar = variable.toStdWString();
    wchar_t *value = _wgetenv(wvar.c_str());
    if (!value) return QString();
    else return QString::fromStdWString(std::wstring(value));
#else
    std::string var = variable.toStdString();
    return QString::fromUtf8(qgetenv(var.c_str()));
#endif
}

static void
putEnvQStr(QString assignment)
{
#ifdef Q_OS_WIN32
    std::wstring wassignment = assignment.toStdWString();
    _wputenv(_wcsdup(wassignment.c_str()));
#else
    putenv(strdup(assignment.toUtf8().data()));
#endif
}

static void
setupTonyVampPath()
{
    QString myVampPath = getEnvQStr("TONY_VAMP_PATH");

#ifdef Q_OS_WIN32
    QChar sep(';');
#else
    QChar sep(':');
#endif
    
    if (myVampPath == "") {
        
        QString appName = QApplication::applicationName();
        QString myDir = QApplication::applicationDirPath();
        QString binaryName = QFileInfo(QCoreApplication::arguments().at(0))
            .fileName();

#ifdef Q_OS_WIN32
        QString programFiles = getEnvQStr("ProgramFiles");
        if (programFiles == "") programFiles = "C:\\Program Files";
        QString pfPath(programFiles + "\\" + appName);
        myVampPath = myDir + sep + pfPath;
#else
#ifdef Q_OS_MAC
        myVampPath = myDir + "/../Resources";
        (void)sep; // unused
#else
        if (binaryName != "") {
            myVampPath =
                myDir + "/../lib/" + binaryName + sep;
        }
        myVampPath = myVampPath +
            myDir + "/../lib/" + appName + sep +
            myDir;
#endif
#endif
    }

    SVCERR << "Setting VAMP_PATH to " << myVampPath
           << " for Tony plugins" << endl;

    QString env = "VAMP_PATH=" + myVampPath;

    // Windows lacks setenv, must use putenv (different arg convention)
    putEnvQStr(env);
}
        
int
main(int argc, char **argv)
{
    if (argc == 2 && (QString(argv[1]) == "--version" ||
                      QString(argv[1]) == "-v")) {
        cerr << TONY_VERSION << endl;
        exit(0);
    }

    svSystemSpecificInitialisation();

    TonyApplication application(argc, argv);

    QApplication::setOrganizationName("sonic-visualiser");
    QApplication::setOrganizationDomain("sonicvisualiser.org");
    QApplication::setApplicationName("Tony");

    setupTonyVampPath();

    QStringList args = application.arguments();

    signal(SIGINT,  signalHandler);
    signal(SIGTERM, signalHandler);

#ifndef Q_OS_WIN32
    signal(SIGHUP,  signalHandler);
    signal(SIGQUIT, signalHandler);
#endif

    bool audioOutput = true;
    bool sonification = true;
    bool spectrogram = true;

    if (args.contains("--help") || args.contains("-h") || args.contains("-?")) {
        std::cerr << QApplication::tr(
            "\nTony is a program for interactive note and pitch analysis and annotation.\n\nUsage:\n\n  %1 [--no-audio] [--no-sonification] [--no-spectrogram] [<file> ...]\n\n  --no-audio: Do not attempt to open an audio output device\n  --no-sonification: Disable sonification of pitch tracks and notes and hide their toggles.\n  --no-spectrogram: Disable spectrogram.\n  <file>: One or more Tony (.ton) and audio files may be provided.").arg(argv[0]).toStdString() << std::endl;
        exit(2);
    }

    if (args.contains("--no-audio")) audioOutput = false;

    if (args.contains("--no-sonification")) sonification = false;

    if (args.contains("--no-spectrogram")) spectrogram = false;

    if (args.contains("--first-run")) {
        QSettings settings;
        settings.clear();
    }

    InteractiveFileFinder::getInstance()->setApplicationSessionExtension("ton");

    QApplication::setAttribute(Qt::AA_UseHighDpiPixmaps);

    QSplashScreen *splash = 0;
    // If we had a splash screen, we would show it here

    QIcon icon;
    int sizes[] = { 16, 22, 24, 32, 48, 64, 128 };
    for (size_t i = 0; i < sizeof(sizes)/sizeof(sizes[0]); ++i) {
        icon.addFile(QString(":icons/tony-%1x%2.png").arg(sizes[i]).arg(sizes[i]));
    }
    QApplication::setWindowIcon(icon);

    QString language = QLocale::system().name();

    QTranslator qtTranslator;
    QString qtTrName = QString("qt_%1").arg(language);
    std::cerr << "Loading " << qtTrName.toStdString() << "..." << std::endl;
    bool success = false;
    if (!(success = qtTranslator.load(qtTrName))) {
        QString qtDir = getenv("QTDIR");
        if (qtDir != "") {
            success = qtTranslator.load
                (qtTrName, QDir(qtDir).filePath("translations"));
        }
    }
    if (!success) {
        std::cerr << "Failed to load Qt translation for locale" << std::endl;
    }
    application.installTranslator(&qtTranslator);

    StoreStartupLocale();
    
    // Permit size_t and PropertyName to be used as args in queued signal calls
    qRegisterMetaType<size_t>("size_t");
    qRegisterMetaType<PropertyContainer::PropertyName>("PropertyContainer::PropertyName");

    MainWindow::AudioMode audioMode = 
        MainWindow::AUDIO_PLAYBACK_NOW_RECORD_LATER;

    if (!audioOutput) {
        audioMode = MainWindow::AUDIO_NONE;
    } 
    
    MainWindow *gui = new MainWindow(audioMode, sonification, spectrogram);
    application.setMainWindow(gui);
    if (splash) {
        QObject::connect(gui, SIGNAL(hideSplash()), splash, SLOT(hide()));
    }

    QScreen *screen = QApplication::primaryScreen();
    QRect available = screen->availableGeometry();

    int width = (available.width() * 2) / 3;
    int height = available.height() / 2;
    if (height < 450) height = (available.height() * 2) / 3;
    if (width > height * 2) width = height * 2;

    QSettings settings;
    settings.beginGroup("MainWindow");
    QSize size = settings.value("size", QSize(width, height)).toSize();
    gui->resizeConstrained(size);
    if (settings.contains("position")) {
        QRect prevrect(settings.value("position").toPoint(), size);
        if (!(available & prevrect).isEmpty()) {
            gui->move(prevrect.topLeft());
        }
    }
    settings.endGroup();
    
    gui->show();

    application.readyForFiles();
    
    for (QStringList::iterator i = args.begin(); i != args.end(); ++i) {

        if (i == args.begin()) continue;
        if (i->startsWith('-')) continue;

        QString path = *i;

        application.handleFilepathArgument(path, splash);
    }

    application.handleQueuedPaths(splash);
        
    if (splash) splash->finish(gui);
    delete splash;

    int rv = application.exec();

    gui->hide();

    cleanupMutex.lock();

    if (!cleanedUp) {
        TransformFactory::deleteInstance();
        TempDirectory::getInstance()->cleanup();
        cleanedUp = true;
    }

    application.releaseMainWindow();

    delete gui;

    cleanupMutex.unlock();

    return rv;
}

/** Application-global handler for filepaths passed in, e.g. as
 * command-line arguments or apple events */

void TonyApplication::handleFilepathArgument(QString path,
                                             QSplashScreen *splash)
{
    static bool haveSession = false;
    static bool haveMainModel = false;
    static bool havePriorCommandLineModel = false;

    if (!m_mainWindow) return;

    MainWindow::FileOpenStatus status = MainWindow::FileOpenFailed;

#ifdef Q_OS_WIN32
    path.replace("\\", "/");
#endif

    if (path.endsWith("ton")) {
        if (!haveSession) {
            status = m_mainWindow->openSessionPath(path);
            if (status == MainWindow::FileOpenSucceeded) {
                haveSession = true;
                haveMainModel = true;
            }
        } else {
            std::cerr << "WARNING: Ignoring additional session file argument \"" << path << "\"" << std::endl;
            status = MainWindow::FileOpenSucceeded;
        }
    }
    if (status != MainWindow::FileOpenSucceeded) {
        if (!haveMainModel) {
            status = m_mainWindow->openPath(path, MainWindow::ReplaceSession);
            if (status == MainWindow::FileOpenSucceeded) {
                haveMainModel = true;
            }
        } else {
            if (haveSession && !havePriorCommandLineModel) {
                status = m_mainWindow->openPath(path, MainWindow::AskUser);
                if (status == MainWindow::FileOpenSucceeded) {
                    havePriorCommandLineModel = true;
                }
            } else {
                status = m_mainWindow->openPath(path, MainWindow::CreateAdditionalModel);
            }
        }
    }
    if (status == MainWindow::FileOpenFailed) {
        if (splash) splash->hide();
        QMessageBox::critical
            (m_mainWindow, QMessageBox::tr("Failed to open file"),
             QMessageBox::tr("File or URL \"%1\" could not be opened").arg(path));
    } else if (status == MainWindow::FileOpenWrongMode) {
        if (splash) splash->hide();
        QMessageBox::critical
            (m_mainWindow, QMessageBox::tr("Failed to open file"),
             QMessageBox::tr("<b>Audio required</b><p>Please load at least one audio file before importing annotation data"));
    }
}

